Add Docker Support

jacob-creech opened the pull request:
PR for nomiclabs/buidler-docker-solc#1

- [x] The plugin should override the internalTask TASK_COMPILE_RUN_COMPILER. You can read its definition here. This is part of PR #190 which will be merged soon.
- [x] The overriden task receives the standard JSON input of solc, and returns the standard output JSON. You can read more about them here.
- [x] The plugin should throw a BuidlerPluginError if docker is not running.
- [x] The plugin should pull the docker with the right version of solc if necessary. The version is specified in the config.
- [x] The plugin should throw a BuidlerPluginError if the version is invalid.
- [x] The plugin should throw a BuidlerPluginError if it can't pull the docker image.
- [x] The plugin should have unit and/or integration tests. We can provide guidance on how to test each part.
- [x] The plugin must be based in the Buidler plugin boilerplate.
- [x] Update README with more documentation
- [x] Update Travis

jacob-creech committed:
Initial Commit

Initial commit for the buidler-docker-solc plugin. Added boilerplate code found at [buidler-ts-plugin-boilerplate](https://github.com/nomiclabs/buidler-ts-plugin-boilerplate). This is related to issue nomiclabs/buidler-docker-solc#1

jacob-creech committed:
Added Overridden TASK_COMPILE_RUN_COMPILER

Override the task TASK_COMPILE_RUN_COMPILER. Added test to run the task. Added docker implementation for everything docker related. Now just to make TASK_COMPILE_RUN_COMPILER run through the docker functions, and add the proper testing. Oh, and of course, add lazy loading to docker containers.

jacob-creech committed:
Updated docker validation

Added validation for versioning

jacob-creech committed:
Refactored docker initialization

Previously, it was possible to get an error with the docker image loading incorrectly displayed because all of the functions were async. Now, loading the docker image does all the validations in order that they have to occur, correctly displaying the correct BuidlerPluginError as necessary

jacob-creech committed:
Use node-docker-api for interacting with docker

Changed out using child_process to using node-docker-api to avoid exposing the rpc. Changed task to be overridden with passing the json for contracts

jacob-creech committed:
Docker run supported

Got docker run with attached volume supported. Adding tests and coverage, plus lazy loading next

jacob-creech committed:
Updated error handling

Docker is not running error was in the incorrect spot, as we'll run into a
 docker error earlier in the flow on requiring docker

jacob-creech committed:
Updated docker configuration

Running solc on your contracts in docker now works.

jacob-creech committed:
Updated test coverage + child_process

Updated docker to run with child_process and return the correct json formatted result from the compiler. Updated test coverage. Updated information under package.json

jacob-creech committed:
Update README.md

Added additional documentation on how to install and use buidler-docker-solc

jacob-creech committed:
Changing test permisions

Updated permissions of the run-tests file so that travis can run the unit tests

jacob-creech committed:
Merge branch 'master' of https://github.com/jacobcreech/buidler-docker-solc

jacob-creech commented:
@alcuadrado Please review and let me know any feedback. I will submit on Gitcoin, but this is still a living PR if further updates are required.

itirabasso commented on src/docker.ts:
```diff
@@ -0,0 +1,57 @@
+import { BuidlerPluginError } from "@nomiclabs/buidler/plugins";
+
+export class Docker {
+  private env: any;
```
This should be a ```BuidlerRuntimeEnvironment```

Also, why is this needed?

itirabasso commented on src/docker.ts:
```diff
@@ -0,0 +1,57 @@
+    this.env = require("@nomiclabs/buidler");
+  }
+
+  public async compile(solcInput: any) {
```
solcInput parameter could be a ```SolcInput```, or at least ```Partial<SolcInput>```

itirabasso commented on src/docker.ts:
```diff
@@ -0,0 +1,57 @@
+  private env: any;
+
+  constructor(private readonly config: any) {
+    this.env = require("@nomiclabs/buidler");
```
This isn't needed.

I'd pass a [SolcConfig](https://buidler.dev/api/interfaces/solcconfig.html) in the constructor (which contains the solc's version) instead of using the whole BuidlerRuntimeEnvironment.

itirabasso replied:
I just noticed that you already have a ```config``` parameter but you don't use it at all.

itirabasso commented on src/docker.ts:
```diff
@@ -0,0 +1,57 @@
+    try {
+      const command =
+        "echo '" +
+        JSON.stringify(solcInput) +
```
I think that using a single apostrophe could lead to some malicious behavior, like a command injection.

I'd use a triple apostrophe to avoid that 
```
const command = "echo '''" +
    JSON.stringify(solcInput) + 
    "''' | docker run -i ethereum/solc:" +
    this.env.config.solc.version +
    " --standard-json";
```

itirabasso commented on package.json:
```diff
@@ -0,0 +1,47 @@
+  },
+  "peerDependencies": {
+    "@nomiclabs/buidler": "^1.0.0-beta.2",
+    "app-root-path": "^2.1.0",
```
I think this isn't used.

itirabasso commented on package.json:
```diff
@@ -0,0 +1,47 @@
+  "peerDependencies": {
+    "@nomiclabs/buidler": "^1.0.0-beta.2",
+    "app-root-path": "^2.1.0",
+    "node-docker-api": "^1.1.22",
```
Idem above

jacob-creech commented:
@itirabasso Thank you for the feedback, I will update asap

jacob-creech committed:
Refactoring for efficiency

Removed unnecessary packages from package. Removed import of entire environment. Passed the solcConfig on Docker object creation, instead of unused config.

jacob-creech committed:
Package bug fix

Package was missing comma
